src/expr/macro/src/gen.rs

```diff
@@ -423,9 +425,188 @@ impl FunctionAttr {
+                    build: #build_fn,
+                    type_infer: #type_infer_fn,
+                }) };
+            }
+        })
+    }
+
+    fn generate_build_table_function(&self) -> Result<TokenStream2> {
+        let num_args = self.args.len();
+        let fn_name = format_ident!("{}", self.user_fn.name);
+        let struct_name = format_ident!("{}", self.ident_name());
+        let ids = (0..num_args)
+            .filter(|i| match &self.prebuild {
+                Some(s) => !s.contains(&format!("${i}")),
+                None => true,
+            })
+            .collect_vec();
+        let const_ids = (0..num_args).filter(|i| match &self.prebuild {
+            Some(s) => s.contains(&format!("${i}")),
+            None => false,
+        });
+        let elems: Vec<_> = ids.iter().map(|i| format_ident!("v{i}")).collect();
+        let all_child: Vec<_> = (0..num_args).map(|i| format_ident!("child{i}")).collect();
+        let const_child: Vec<_> = const_ids.map(|i| format_ident!("child{i}")).collect();
+        let child: Vec<_> = ids.iter().map(|i| format_ident!("child{i}")).collect();
+        let array_refs: Vec<_> = ids.iter().map(|i| format_ident!("array{i}")).collect();
+        let arrays: Vec<_> = ids.iter().map(|i| format_ident!("a{i}")).collect();
+        let arg_arrays = self
+            .args
+            .iter()
+            .map(|t| format_ident!("{}", types::array_type(t)));
+        let array_builder = format_ident!("{}Builder", types::array_type(&self.ret));
+        let const_arg = match &self.prebuild {
+            Some(_) => quote! { &self.const_arg },
+            None => quote! {},
+        };
+        let const_arg_type = match &self.prebuild {
+            Some(s) => s.split("::").next().unwrap().parse().unwrap(),
+            None => quote! { () },
+        };
+        let const_arg_value = match &self.prebuild {
+            Some(s) => s
+                .replace('$', "child")
+                .parse()
+                .expect("invalid prebuild syntax"),
+            None => quote! { () },
+        };
+        let value = match self.user_fn.return_type {
+            ReturnType::T => quote! { Some(value) },
+            ReturnType::Option => quote! { value },
+            ReturnType::Result => quote! { Some(value?) },
+            ReturnType::ResultOption => quote! { value? },
+        };
+
+        Ok(quote! {
+            |return_type, chunk_size, children| {
+                use risingwave_common::array::*;
+                use risingwave_common::types::*;
+                use risingwave_common::buffer::Bitmap;
+                use risingwave_common::util::iter_util::ZipEqFast;
+                use itertools::multizip;
+
+                crate::ensure!(children.len() == #num_args);
+                let mut iter = children.into_iter();
+                #(let #all_child = iter.next().unwrap();)*
+                #(let #const_child = #const_child.eval_const()?;)*
```

Comment:
Are we planning to support non-constant pre-build parameters? It may perform worse but can be more general.

Reply:
Sure. There is such a requirement in scalar functions `regexp_match`, `to_char` and `to_timestamp`.